specify/specify7
# initial schema creation for geospecify

Fixes #4981
Fixes #4827
Fixes #4976

This is the initial schema create for geo-specify, all in one Django migration.  Right now the new models are being added to the `sp7_models` Django app, but we might be able to change that soon and use the `specify` app instead.  This PR adds the `CollectionObjectType`, `CollectionObjectGroup`, and `CollectionObjectGroupJoin` models to the Specify 7 schema.  It also modifies the `CollectionObject` model by adding two new fields.

This PR also magically creates an initial Django migration for all the inherited schema from Specify 6.  This is done  by editing the django_migrations table with an SQL statement that is triggered in the docker entry point file.

![image](https://github.com/specify/specify7/assets/2774057/e2916def-9f1e-478e-b7f9-c344cc5ff94e)

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

When switch to the branch, run `docker compose down` and `docker compose up --build`

For testing on the test panel, if there are no errors when starting up the instance then the migration ran without any problems.  Check to make sure that the follow tables exist in the QB:

- [x] CollectionObjectType - https://hujinnhcgeo-geodude.test.specifysystems.org/specify/query/new/collectionobjecttype/
- [x] CollectionObjectGroupType - https://hujinnhcgeo-geodude.test.specifysystems.org/specify/query/new/collectionobjectgrouptype/
- [x] CollectionObjectGroup - https://hujinnhcgeo-geodude.test.specifysystems.org/specify/query/new/collectionobjectgroup/
- [x] CollectionObjectGroupJoin - https://hujinnhcgeo-geodude.test.specifysystems.org/specify/query/new/collectionobjectgroupjoin/

Verify that each Collection has a default CollectionObjectType
- [x] Verify each Collection has its CollectionObjectType pointing to something, so they are not null.  Do this by doing a query on all collections and verify that each has something in its CollectionObjectType field.  Each CollectionObjectType name should match the name of the discipline associated with the collection.  You'll have to go through each discipline to verify this for all the collections unfortunately.

Before testing the new Trees, generate them by calling the `/api/specify_trees/add_geo_default_trees/` API call.  There is no front-end way of calling this API at the moment, so it will need to be done with a curl command.  Message me when testing and I can run the command for that test panel instance.  (Maybe we can wait to test this in a later PR where the front-end is implemented)
EDIT: You can now run this command on the test-panel for your container instance that will create the default geology trees `docker exec -it <container_name> ve/bin/python manage.py create_geo_trees`.  For the database I'll set up on the test panel, this step should already be done.

Verify default geo trees (cannot currently be tested with the UI):
- [x] Default CollectionObjectType records matching each discipline
- [x] Default TaxonTreeDefs created for geo.
- [x] Default TaxonTreeDefItems (Tree Ranks) created for each of the geo trees.

Note: Might not be able to verify the default in the current UI, might have to check the database.

Testing on local docker instance (only if you have specify docker setup):
- The migration should automatically happen when starting your docker instance.
- To manually run the migration first run `ve/bin/python manage.py base_specify_migration`
- Then run `ve/bin/python manage.py migrate specify 0002_geo`
- To test that reverting the migration works run `ve/bin/python manage.py migrate specify 0001_initial`
- Look through the database to verify that the migration created the CollectionObjectType, CollectionObjectGroup, CollectionObjectGroupJoin tables, as well as the default CollectionObjectType, TaxonTreeDef, and TaxonTreeDefItem records.
